cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java

```diff
@@ -1,195 +1,213 @@
 
   @GuardedBy("lock")
   private final Map<T, Snapshot> snapshots = new HashMap<>();
   @GuardedBy("lock")
-  private final Table<T, Long, Watch> watches = HashBasedTable.create();
+  private final Map<T, CacheStatusInfo> statuses = new HashMap<>();
 
   @GuardedBy("lock")
   private long watchCount;
 
-  /**
-   * Constructs a simple cache that uses {@link ForkJoinPool#commonPool()} to execute async callbacks.
-   *
-   * @param callback callback invoked when a group is seen for the first time
-   * @param groups maps an envoy host to a node group
-   */
-  public SimpleCache(Consumer<T> callback, NodeGroup<T> groups) {
-    this(callback, groups, ForkJoinPool.commonPool());
-  }
-
   /**
    * Constructs a simple cache.
    *
-   * @param callback callback invoked when a group is seen for the first time
+   * @param ads if running in ADS-mode, responses are delayed until all resources are explicitly named
    * @param groups maps an envoy host to a node group
-   * @param executorService executor service used to execute async callbacks
    */
-  public SimpleCache(Consumer<T> callback, NodeGroup<T> groups, ExecutorService executorService) {
-    this.callback = callback;
+  public SimpleCache(boolean ads, NodeGroup<T> groups) {
+    this.ads = ads;
     this.groups = groups;
-    this.executorService = executorService;
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
-  public void setSnapshot(T group, Snapshot snapshot) {
-    synchronized (lock) {
-      // Update the existing entry.
-      snapshots.put(group, snapshot);
+  public Watch createWatch(DiscoveryRequest request) {
+    T group = groups.hash(request.getNode());
 
-      // Trigger existing watches
-      Set<Long> idsToRemove = watches.row(group).entrySet().stream()
-          .map(watchEntry -> {
-            respond(watchEntry.getValue(), snapshot, group);
-            return watchEntry.getKey();
-          })
-          .collect(Collectors.toSet());
-
-      // Discard all watches; the client must request a new watch to receive updates and ACK/NACK.
-      //
-      // Note that the removals have to happen after we're done iterating over the row Map otherwise
-      // we get a java.util.ConcurrentModificationException
-      idsToRemove.forEach(idToRemove -> watches.remove(group, idToRemove));
-    }
-  }
-
-  /**
-   * {@inheritDoc}
-   */
-  @Override
-  public Watch watch(ResourceType type, Node node, String version, Collection<String> names) {
-    Watch watch = new Watch(names, type);
+    writeLock.lock();
 
-    final T group;
-
-    // Do nothing if group hashing failed.
     try {
-      group = groups.hash(node);
-    } catch (Exception e) {
-      LOGGER.error("failed to hash node {} into group", node, e);
+      CacheStatusInfo status = statuses.computeIfAbsent(group, g -> new CacheStatusInfo(request.getNode()));
 
-      return watch;
-    }
+      status.setLastWatchRequestTime(System.currentTimeMillis());
```

Comment:
are we taking `writeLock` twice? one here and one inside `setLastWatchRequestTime` of `CacheStatusInfo`?

Reply:
nope, those are two different locks. the one here is owned by `SimpleCache` and the one instead `setLastWatchRequestTime` is owned by the `CacheStatusInfo`.

Reply:
ok. I see. I did not notice that statuses is a simple hash map. So this lock here is to protect concurrent inserts to that map. I was thinking nothing is being updated here that is why thinking why we need lock here.